pkg/virt-handler/vm.go

```diff
@@ -3005,150 +2999,246 @@ func (c *VirtualMachineController) vmUpdateHelperDefault(origVMI *v1.VirtualMach
-			}
-			if err := diskutils.DefaultOwnershipManager.SetFileOwnership(sevDevice); err != nil {
-				return fmt.Errorf("failed to set SEV device owner: %v", err)
-			}
-		}
+	isolationRes, err := c.podIsolationDetector.Detect(vmi)
+	if err != nil {
+		return fmt.Errorf(failedDetectIsolationFmt, err)
+	}
 
-		if virtutil.IsNonRootVMI(vmi) {
-			if err := c.nonRootSetup(origVMI); err != nil {
-				return err
-			}
-		}
-		for _, fs := range vmi.Spec.Domain.Devices.Filesystems {
-			socketPath, err := isolation.SafeJoin(isolationRes, virtiofs.VirtioFSSocketPath(fs.Name))
-			if err != nil {
-				return err
-			}
-			if err := diskutils.DefaultOwnershipManager.SetFileOwnership(socketPath); err != nil {
-				return err
-			}
-		}
+	if err := c.downwardMetricsManager.StartServer(vmi, isolationRes.Pid()); err != nil {
+		return err
+	}
 
-		// set runtime limits as needed
-		err = c.podIsolationDetector.AdjustResources(vmi, c.clusterConfig.GetConfig().AdditionalGuestMemoryOverheadRatio)
+	if err := c.setupNetwork(vmi, netsetup.FilterNetsForLiveUpdate(vmi)); err != nil {
+		log.Log.Object(vmi).Error(err.Error())
+		c.recorder.Event(vmi, k8sv1.EventTypeWarning, "NicHotplug", err.Error())
+		*errorTolerantFeaturesError = append(*errorTolerantFeaturesError, err)
+	}
+
+	return nil
+}
+
+// handleStartingVMI: Contains the logic for starting VMs (container disks, initial network setup, device ownership).
+func (c *VirtualMachineController) handleStartingVMI(
+	vmi *v1.VirtualMachineInstance,
+	cgroupManager cgroup.Manager,
+	disksInfo *map[string]*containerdisk.DiskInfo,
+) (bool, error) {
+	// give containerDisks some time to become ready before throwing errors on retries
+	info := c.getLauncherClientInfo(vmi)
+	if ready, err := c.containerDiskMounter.ContainerDisksReady(vmi, info.NotInitializedSince); !ready {
 		if err != nil {
-			return fmt.Errorf("failed to adjust resources: %v", err)
+			return false, err
 		}
+		c.queue.AddAfter(controller.VirtualMachineInstanceKey(vmi), time.Second*1)
+		return false, nil
+	}
 
-		if util.IsSEVAttestationRequested(vmi) {
-			sev := vmi.Spec.Domain.LaunchSecurity.SEV
-			if sev.Session == "" || sev.DHCert == "" {
-				// Wait for the session parameters to be provided
-				return nil
-			}
-		}
-	} else if vmi.IsRunning() {
-		if err := c.hotplugSriovInterfaces(vmi); err != nil {
-			log.Log.Object(vmi).Error(err.Error())
+	var err error
+	*disksInfo, err = c.containerDiskMounter.MountAndVerify(vmi)
+	if err != nil {
+		return false, err
+	}
+
+	if err := c.hotplugVolumeMounter.Mount(vmi, cgroupManager); err != nil {
+		return false, err
+	}
+
+	if err := c.setupNetwork(vmi, netsetup.FilterNetsForVMStartup(vmi)); err != nil {
+		return false, fmt.Errorf("failed to configure vmi network: %w", err)
+	}
+
+	isolationRes, err := c.podIsolationDetector.Detect(vmi)
+	if err != nil {
+		return false, fmt.Errorf(failedDetectIsolationFmt, err)
+	}
+
+	if err := c.setupDevicesOwnerships(vmi, isolationRes); err != nil {
```

Comment:
Hi @vladikr, sorry to jump in late but shouldn't we be doing the same on the target node on migration? Right now we setup different devices on-startup vs on-migration.